src/algorithm/neighbour/fastpair.rs

```diff
@@ -0,0 +1,554 @@
+#![allow(non_snake_case)]
+use itertools::Itertools;
+///
+/// FastPair: Data-structure for the dynamic closest-pair problem.
+///
+/// Reference:
+///  Eppstein, David: Fast hierarchical clustering and other applications of
+///  dynamic closest pairs. Journal of Experimental Algorithmics 5 (2000) 1.
+///
+use std::collections::HashMap;
+
+use crate::algorithm::neighbour::dissimilarities::PairwiseDissimilarity;
+use crate::error::{Failed, FailedError};
+use crate::linalg::Matrix;
+use crate::math::distance::euclidian::Euclidian;
+use crate::math::num::RealNumber;
+
+///
+/// FastPair
+///
+/// Ported from Python implementation:
+/// <https://github.com/carsonfarmer/fastpair/blob/b8b4d3000ab6f795a878936667eee1b557bf353d/fastpair/base.py>
+/// MIT License (MIT) Copyright (c) 2016 Carson Farmer
+///
+/// affinity used is Euclidean so to allow linkage with single, ward, complete and average
+///
+#[derive(Debug, Clone)]
+pub struct FastPair<'a, T: RealNumber, M: Matrix<T>> {
+    /// initial matrix
+    samples: &'a M,
+    /// closest pair hashmap (connectivity matrix for closest pairs)
+    pub distances: HashMap<usize, PairwiseDissimilarity<T>>,
+    /// conga line used to keep track of the closest pair
+    pub neighbours: Vec<usize>,
+}
+
+impl<'a, T: RealNumber, M: Matrix<T>> FastPair<'a, T, M> {
+    ///
+    /// Constructor
+    /// Instantiate and inizialise the algorithm
+    ///
+    pub fn new(m: &'a M) -> Result<Self, Failed> {
```

Comment:
Not sure if we should use `new` or `fit`?  Any suggestions @VolodymyrOrlov ?

Reply:
This implementation is meant to be a lower level tool that is not exposed as a public interface. At some point we will have something high-level that use this algorithm, at that point the higher-level implementation will have a `fit` method that will instantiate `FastPair::new`.
Applying `fit` to `FastPair` doesn't make sense as it is not a regression or any other ML implementation but an algorithm that can be reused in many ML implementations.